api/src/main/java/org/apache/iceberg/io/FileReader.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import java.util.Map;
+
+/**
+ * A reader capable of reading files and returning records of type {@code D}.
+ *
+ * <p>This interface should be implemented by classes that not only read records from files but also
+ * provide metadata stored in the file.
+ *
+ * @param <D> the record type
+ */
+public interface FileReader<D> extends CloseableIterable<D> {
+  Map<String, String> meta();
```

Comment:
I don't think that we should add this and should instead track down and remove cases where we are using the file metadata to recover a partition spec. I think that only happens for unused and deprecated calls in public methods exposed in the core module.

Reply:
Reverted.
CC: @RussellSpitzer, @danielcweeks